src/source/CurlApiCallbacks.c

```diff
@@ -1896,25 +2025,83 @@ STATUS tagResourceCachingCurl(UINT64 customData, PCHAR streamArn, UINT32 tagCoun
                               PServiceCallContext pServiceCallContext)
 {
     ENTERS();
-    UNUSED_PARAM(streamArn);
-    UNUSED_PARAM(tagCount);
-    UNUSED_PARAM(tags);
-
     STATUS retStatus = STATUS_SUCCESS;
     PCurlApiCallbacks pCurlApiCallbacks = (PCurlApiCallbacks) customData;
     PCallbacksProvider pCallbacksProvider = NULL;
     STREAM_HANDLE streamHandle;
+    BOOL endpointsLocked = FALSE, emulateApiCall = TRUE;
+    UINT64 curTime, value;
+    PEndpointTracker pEndpointTracker = NULL;
 
     CHK(pCurlApiCallbacks != NULL && pCurlApiCallbacks->pCallbacksProvider != NULL && pServiceCallContext != NULL, STATUS_INVALID_ARG);
     pCallbacksProvider = pCurlApiCallbacks->pCallbacksProvider;
 
     streamHandle = (STREAM_HANDLE) pServiceCallContext->customData;
-    retStatus = tagResourceResultEvent(streamHandle, SERVICE_CALL_RESULT_OK);
 
+    // We check whether we have already made the call by checking
+    // for the presence of the endpoint in the cache.
+    switch(pCurlApiCallbacks->cacheType) {
+        case API_CALL_CACHE_TYPE_NONE:
+            emulateApiCall = FALSE;
+            break;
+
+        case API_CALL_CACHE_TYPE_ENDPOINT_ONLY:
+            emulateApiCall = TRUE;
+            break;
+
+        case API_CALL_CACHE_TYPE_ALL:
+            pCallbacksProvider->clientCallbacks.lockMutexFn(pCallbacksProvider->clientCallbacks.customData,
+                                                            pCurlApiCallbacks->cachedEndpointsLock);
+            endpointsLocked = TRUE;
+
+            // Attempt to retrieve the cached value
+            retStatus = hashTableGet(pCurlApiCallbacks->pCachedEndpoints, (UINT64) streamHandle, &value);
+
+            CHK(retStatus == STATUS_HASH_KEY_NOT_PRESENT || retStatus == STATUS_SUCCESS, retStatus);
+
+            if (retStatus == STATUS_HASH_KEY_NOT_PRESENT) {
+                emulateApiCall = FALSE;
+
+                // Reset the status if not found
+                retStatus = STATUS_SUCCESS;
+            } else {
+                pEndpointTracker = (PEndpointTracker) value;
+                curTime = pCallbacksProvider->clientCallbacks.getCurrentTimeFn(pCallbacksProvider->clientCallbacks.customData);
+
+                if (pEndpointTracker == NULL ||

```

Comment:
if cache has expired, tag resource wont get new endpoint.

Reply:
We use the presence of the endpoint in the cache as an indicator whether the other calls have been made or not as the state machine is pretty much sequential. Get endpoint is the last in the sequence of FE API calls. Next API call is the actual PutMedia call